sensor/mcp9808/mcp9808.go

```diff
@@ -0,0 +1,484 @@
+func (d *MCP9808) SetAlertPolarity(set bool) {
+	d.flipConfigBit(configAlertPolarity, set)
+}
+
+// AlertMode - Alert Output Mode bit
+// 0 (false) = Comparator output (power-up default)
+// 1 (true) = Interrupt output
+func (d *MCP9808) AlertMode() bool {
+	return d.readConfigValue(configAlertMode)
+}
+
+// SetAlertMode - Alert Output Mode bit
+// 0 (false) = Comparator output (power-up default)
+// 1 (true) = Interrupt output
+// This bit cannot be altered when either of the Lock bits are set (bit 6 and bit 7).
+// This bit can be programmed in Shutdown mode, but the Alert output will not assert or deassert.
+func (d *MCP9808) SetAlertMode(set bool) {
+	d.flipConfigBit(configAlertMode, set)
+}
+
+// Temperature contains the ambient temperature along with alert values.
+type Temperature struct {
+	CelsiusDeg                            float64
+	AboveCritical, AboveUpper, BelowLower bool
+}
+
+// AmbientTemp reads the current sensor value along with the flags denoting what boundaries the
+// current temperature exceeds.
+func (d *MCP9808) AmbientTemp() (*Temperature, error) {
+	temp, err := d.Bus.ReadWordFromReg(address, regAmbientTemp)
+	if err != nil {
+		return nil, err
+	}
+
+	tempC := temp &^ 0xF000
+
+	wholeNum := float64(tempC&^0xF) / 16.0
+	fraction := tempC &^ 0xFF0
+
+	tempResult := &Temperature{
+		AboveCritical: !(temp&(1<<15) == 0),
+		AboveUpper:    !(temp&(1<<14) == 0),
+		BelowLower:    !(temp&(1<<13) == 0)}
+	tempResult.CelsiusDeg = wholeNum + (float64(fraction) / 10000.0)
+	if !(temp&(1<<12) == 0) { // read sign bit
+		tempResult.CelsiusDeg *= -1
+	}
+	return tempResult, nil
+}
+
+// readTempC reads from the ambient temperature register and returns the current temperature value in celsius
+func (d *MCP9808) readTempC(reg byte) (float64, error) {
+	temp, err := d.Bus.ReadWordFromReg(address, reg)
+	if err != nil {
+		return 0, err
+	}
+
+	log.Printf("\nTA vs TCrit: %v\nTA vs TUpper: %v\nTA vs TLower: %v\n", !(temp&(1<<15) == 0), !(temp&(1<<14) == 0), !(temp&(1<<13) == 0))
+
+	wholeTempC := temp &^ 0xF000
+
+	wholeNum := float64(wholeTempC&^0xF) / 16.0
+	fraction := wholeTempC &^ 0xFF0
+
+	tempRead := wholeNum + (float64(fraction) / 10000.0)
+	if !(temp&(1<<12) == 0) { // read sign bit
+		tempRead *= -1
+	}
+	return tempRead, nil
+}
+
+func (d *MCP9808) setTemp(reg byte, newTemp float64) error {
+	var signBit uint16
+	if newTemp < 0 {
+		newTemp *= -1
+		signBit = 0x1000
+	}
+	newTempWord := uint16(newTemp)*15 + uint16(newTemp*16) + signBit
+
+	if err := d.Bus.WriteWordToReg(address, reg, newTempWord); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CriticalTempUpper reads the current temperature set in the critical temperature register.
+func (d *MCP9808) CriticalTempUpper() (float64, error) {
+	return d.readTempC(regCriticalTemp)
+}
+
+// SetCriticalTemp when the temperature goes above the set value the alert will be
+// triggered if enabled.
+func (d *MCP9808) SetCriticalTemp(newTemp float64) error {
```

Comment:
Most of the other Set functions are writing to the config which I thought would be cleaner to make edits locally and then once you are satisfied with your config changes you can write them to the device.